staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go

```diff
@@ -217,28 +269,48 @@ func (s *openAPIAggregator) handlerWithUser(handler http.Handler, info user.Info
 	})
 }
 
+func etagFor(data []byte) string {
+	return fmt.Sprintf("\"%X\"", sha512.Sum512(data))
+}
+
 // downloadOpenAPISpec downloads openAPI spec from /swagger.json endpoint of the given handler.
-func (s *openAPIAggregator) downloadOpenAPISpec(handler http.Handler) (*spec.Swagger, error) {
+// httpStatus is only valid if err == nil
+func (s *openAPIAggregator) downloadOpenAPISpec(handler http.Handler, etag string) (returnSpec *spec.Swagger, newEtag string, httpStatus int, err error) {
 	handler = s.handlerWithUser(handler, &user.DefaultInfo{Name: aggregatorUser})
 	handler = request.WithRequestContext(handler, s.contextMapper)
 	handler = http.TimeoutHandler(handler, specDownloadTimeout, "request timed out")
 
 	req, err := http.NewRequest("GET", "/swagger.json", nil)
 	if err != nil {
-		return nil, err
+		return nil, "", 0, err
+	}
+	if len(etag) > 0 {
+		req.Header.Add("If-None-Match", etag)
 	}
 	writer := newInMemoryResponseWriter()
 	handler.ServeHTTP(writer, req)
 
 	switch writer.respCode {
+	case http.StatusNotModified:
+		if len(etag) == 0 {
+			return nil, etag, http.StatusNotModified, fmt.Errorf("http.StatusNotModified is not allowed in absense of etag")
+		}
+		return nil, etag, http.StatusNotModified, nil
+	case http.StatusNotFound:
+		// Gracefully skip 404, assuming the server won't provide any spec
+		return nil, "", http.StatusNotFound, nil
 	case http.StatusOK:
 		openApiSpec := &spec.Swagger{}
 		if err := json.Unmarshal(writer.data, openApiSpec); err != nil {
-			return nil, err
+			return nil, "", 0, err
+		}
+		etag = writer.Header().Get("Etag")
+		if len(etag) == 0 {
+			etag = etagFor(writer.data)
 		}
-		return openApiSpec, nil
+		return openApiSpec, etag, http.StatusOK, nil
 	default:
-		return nil, fmt.Errorf("failed to retrive openAPI spec, http error: %s", writer.String())
+		return nil, "", 0, fmt.Errorf("failed to retrive openAPI spec, http error: %s", writer.String())
```

Comment:
I've commented on the function to say the httpCode is only valid if err == nil. We don't want to confuse the caller by providing both err and httpCode giving them the impression that httpCode can be reliable even if err is not nil. The caller should always check for err first.